### PR TITLE
support cypher parameter(variable)

### PR DESCRIPTION
**Pull request by czpmango**
#### What type of PR is this?
- [ ] bug
- [x] feature
- [ ] enhancement

#### What does this PR do?
Support cypher parameter with ngql variable.

#### Which issue(s)/PR(s) this PR relates to?
#2593 

#### Checklist：
- [x] Documentation affected （Please add the label if documentation needs to be modified.)
- [ ] Incompatible （If it is incompatible, please describe it and add corresponding label.）
- [ ] Need to cherry-pick （If need to cherry-pick to some branches, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

#### Release notes：
Support for parameterized statement must be added to the release notes.

### CODE REVIEW

File: src/graph/executor/query/AppendVerticesExecutor.cpp [hunk 1]
```diff
@@ -45,7 +45,7 @@ folly::Future<Status> AppendVerticesExecutor::appendVertices() {
                  av->exprs(),
                  av->dedup(),
                  av->orderBy(),
-                 av->limit(),
+                 av->limit(qctx()),
```
Comment (Shylock-Hg) [inline src/graph/executor/query/AppendVerticesExecutor.cpp:48]: Why change this?

File: src/graph/planner/plan/Query.h [hunk 2]
```diff
@@ -801,10 +799,9 @@ class Sample final : public SingleInputNode {
   }
 
   // Get constant count
-  int64_t count() const {
-    DCHECK(ExpressionUtils::isEvaluableExpr(count_));
-    QueryExpressionContext qec;
-    auto count = count_->eval(qec).getInt();
+  int64_t count(const QueryContext* qctx) const {
+    DCHECK(ExpressionUtils::isEvaluableExpr(count_, qctx));
+    auto count = count_->eval(QueryExpressionContext(qctx_->ectx())()).getInt();
```
Comment (Shylock-Hg) [inline src/graph/planner/plan/Query.h:804]: Why change this?

Comment (czpmango) [reply src/graph/planner/plan/Query.h:804]: You are right, there is no need to support stateful checking on `Sample` plan node, I will change it back.

File: src/graph/service/GraphService.cpp [hunk 3]
```diff
@@ -114,6 +114,48 @@ void GraphService::signout(int64_t sessionId) {
   sessionManager_->removeSession(sessionId);
 }
 
+folly::Future<ExecutionResponse> GraphService::future_executeWithParameter(
+    int64_t sessionId,
+    const std::string& query,
+    const std::unordered_map<std::string, Value>& parameterMap) {
```
Comment (jievince) [inline src/graph/service/GraphService.cpp:120]: Could we unify `future_executeWithParameter` and `future_execute`? They share almost the same code.

File: src/graph/visitor/EvaluableExprVisitor.cpp [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+EvaluableExprVisitor::EvaluableExprVisitor(const QueryContext *qctx) : qctx_(qctx) {}
+
+void EvaluableExprVisitor::visit(ConstantExpression *) { isEvaluable_ = true; }
+
+void EvaluableExprVisitor::visit(LabelExpression *) { isEvaluable_ = false; }
+
+void EvaluableExprVisitor::visit(UUIDExpression *) { isEvaluable_ = false; }
+
+void EvaluableExprVisitor::visit(VariableExpression *expr) {
+  isEvaluable_ = (qctx_ && qctx_->existParameter(expr->var())) ? true : false;
```
Comment (jievince) [inline src/graph/visitor/EvaluableExprVisitor.cpp:22]:
```
DCHECK(qctx_)
isEvaluable_ = (qctx_->existParameter(expr->var())) ? true : false;
```

File: src/parser/parser.yy [hunk 5]
```diff
@@ -939,7 +939,12 @@ vertex_prop_expression
 
 var_prop_expression
     : VARIABLE DOT name_label {
-        $$ = VariablePropertyExpression::make(qctx->objPool(), *$1, *$3);
+        if (!qctx->existParameter(*$1)) {
+            $$ = VariablePropertyExpression::make(qctx->objPool(), *$1, *$3);
+        } else {
+            $$ = AttributeExpression::make(qctx->objPool(),
```
Comment (jievince) [inline src/parser/parser.yy:945]: Why need `AttributeExpression` here?

File: tests/Makefile [hunk 6]
```diff
@@ -13,7 +13,7 @@ RM_DIR ?= true
 TEST_DIR ?= $(CURR_DIR)
 BUILD_DIR ?= $(CURR_DIR)/../build
 DEBUG ?= true
-J ?= 10
+J ?= 1
```
Comment (jievince) [inline tests/Makefile:16]: Don't forget to change it back

**Review by jievince — COMMENTED** [review]
If we really decide to unify variables and parameters, so this PR is ok to me.

File: src/parser/parser.yy (hunk 5 above)
Comment (czpmango) [reply src/parser/parser.yy:952]: In order to achieve expression like `map.k1.k2.k3`. 
IMO, `VariablePropertyExpression` can also be implemented as `VariableExpression`+`AttributeExpression`.

File: src/graph/service/GraphService.cpp (hunk 3 above)
Comment (czpmango) [reply src/graph/service/GraphService.cpp:120]: After nebula-python's pr been merged.

File: src/graph/visitor/EvaluableExprVisitor.cpp (hunk 4 above)
Comment (czpmango) [reply src/graph/visitor/EvaluableExprVisitor.cpp:22]: `qctx_` is not necessary in the implementation of ngql. Just to be compatible with the previous implementation, the developer can control the intent of the EvaluableExprVisitor.

File: tests/tck/features/yield/parameter.feature [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+  Scenario: return parameters
+    When executing query:
+      """
+      RETURN abs($p1)+1 AS ival, $p2 and false AS bval, $p3+"ef" AS sval, round($p4)+1.1 AS fval, $p5 AS lval, $p6.a AS mval, all(item in $p7.a.b.d where item<4 or ((item>0) is null)) AS pval
+      """
+    Then the result should be, in any order:
+      | ival | bval  | sval           | fval | lval       | mval | pval |
+      | 2    | false | "Tim Duncanef" | 4.1  | [1,true,3] | 3    | true |
+
+  Scenario: cypher with parameters
```
Comment (CPWstatic) [inline tests/tck/features/yield/parameter.feature:19]: We should add some test cases like:
```
$var = ...;
Match ... Where $var Return ...;
```

Comment (czpmango) [reply tests/tck/features/yield/parameter.feature:19]: Added.

File: src/parser/parser.yy [hunk 8]
```diff
@@ -1330,6 +1341,15 @@ vid
         $$ = ConstantExpression::make(qctx->objPool(), *$1);
         delete $1;
     }
+    | VARIABLE {
+        $$ = nullptr;
+        UNUSED($1);
```
[the remark this answers is not in the document]
Comment (czpmango) [reply src/parser/parser.yy]: Fixed.

**Review by CPWstatic — APPROVED** [review]

**Review by jievince — APPROVED** [review]

**Pull request merged**